py4web/utils/url_signer.py

```diff
@@ -10,66 +10,64 @@ class URLVerifier(Fixture):
      to action.uses() to check for the validity of signatures, and the
      sign() method can be used to sign a URL.  If an object of this class
      is passed to the URL helper, it can be used to sign a URL."""
-    def __init__(self, url_signer, query_keys):
```

Comment:
The reason for this was the following:

<a onclick="axios.get('[[=URL('xyz', vars=dict(x=1, y=2), signer=signer)]]'+'&z='+windows.data.z)">click me</a>

Here x,y are signed. z is not because generated in JS. How do we communicate this to the action? Should we not allow this? Maybe not, but why not?

Reply:
Yes, and you can still do the same.   The list of keys to be signed is part of the url_signer, that is passed as an argument, so there is no need to give separately the list of query_keys to be signed.

Reply:
The philosophical sticky point this: do we want one signer per app or one signer per url? I would prefer one per app because we could, for example, create it by default in Auth, but this requires that some parameters be passed not to constructor but to verify(). If we decide we want one signer per action, then there is reason for verify() and/or two classes. Yet this would be a very unusual fixture because we would need one instance of the fixture for each signed action.

Reply:
Let me argue two things: (a) one almost never needs to sign variables, and so in practice only one signer per app is enough, and (b) even if you need to sign variables, at that point, multiple signers is the best.  This means, the current implementation is the correct one. 

This is of course my opinion, but here is my argument. 

(a) signing variables is generally not necessary. 

This is true for two reasons.  First, if one signs a URL u?v=a, then one expects v to be present also in the subsequent use of the URL, and so, one would use as URL u/a , without the variable v.  That is, variables are generally used for optional parameters, not compulsory ones, and the very fact of signing the URL inclusive of the variable shows that the variable in fact is compulsory.  

Second, the main reason for signing URLs is as AJAX endpoints, and it is much easier to give an endpoint of u or u/a, rather than one of u?v=a, because an endpoint without variables (such as u or u/a) is easier to use in Javascript by adding additional variables as in GET calls. 

Indeed, in my own development and in the classes I taught, the old web2py default of including variables in the signature and verification was a constant cause of headaches. 

(b) even if one wants to include variables in the signature, it is much better to then use multiple signers. 

In fact, suppose you have only one signer per app.  Then to sign you would need something like: 

```
signed_url = URL('/something', vars=dict(a=2, b=3), url_signer=url_signer, signed_vars=["a"])
```

and then to verify: 

```
@action.uses(url_signer.verify(signed_vars=["a"]))
```

This is (in my opinion) simply horrible, as there is a list that you must repeat identically in two different places, a big NONO in my mind.  Much better to create a `url_signer_with_a` once and forall, and use that both in the URL constructor, and in the @action.uses method. 

Of course, this is just my opinion.

Reply:
OK. I am convinced. :-)